### PR TITLE
Engines

### PR DESCRIPTION
**Pull request by daanelson**
WIP: Refactor of llama inference into separate engines which can serve predictions with different underlying inference code. 

This has: 
- proper config and predict.py setup (note - we can *definitely* really clean up our config here, pretty sure it can become a ~50 line .yaml file per model at most)
- code refactored into `engine` modules

This does not yet have: 
- per-engine dependency management 
- the aforementioned clean config
- more than one engine :D

### CODE REVIEW

**Comment by daanelson** [comment]
one thing that's coming out of this, at least imo, is that a lot of the config we're setting really lives at the engine level - that's especially true if we have special requirements for `cog.yaml` per engine, etc.

File: model_templates/config.py [hunk 1]
```diff
@@ -24,24 +24,28 @@
...
-USE_EXLLAMA_FOR_UNTRAINED_WEIGHTS = 
 
 
+# ENGINE CONFIGURATION
+# -------------------------------
+# Here we define the specific inference engine we intend to use for inference, and all appropriate kwargs. 
+# -------------------------------
+
+
+ENGINE =
```
Comment (jakedahn) [inline model_templates/config.py:46]: this operator needs a value

Comment (jakedahn) [reply model_templates/config.py:46]: er wait, maybe this is intentionally blank/invalid, perhaps ignore :D

File: predict.py [hunk 2]
```diff
@@ -48,21 +42,22 @@ def setup(self, weights: Optional[Path] = None):
...
+
+        self.engine = ENGINE(base_weights, **ENGINE_KWARGS)
 
         if weights is not None and weights.name == "weights":
             # bugfix
             weights = None
         if weights:
             # If weights are passed in, they are LoRa weights
             # so we need to download the fp16 weights and load with peft
+            # TODO: pass engine and weights both in here.
```
Comment (jakedahn) [inline predict.py:60]: curious what your thoughts on what this should look like?

like should it be `self.engine.initialize_peft(weights)`, or some other approach?

Comment (jakedahn) [inline predict.py:60]: is this todo done? looks like we've got `self.engine.load_lora` below

File: src/inference_engines/__init__.py
Comment (moinnadeem) [inline src/inference_engines/__init__.py:1]: Is it worth having an engines registry of sort that we can pull from?

Comment (daanelson) [reply src/inference_engines/__init__.py:1]: sounds interesting. do you mean adding all of our engines to `__init__` (very in favor of doing this) or something else?

File: predict.py (hunk 2 above)
Comment (daanelson) [reply predict.py:61]: yup, will strike it.

**Comment by technillogue** [comment]
are we still going to merge this?

**Comment by technillogue** [comment]
replaced by #47

**Pull request closed without merging**